### DISCUSSION TITLE
Unifying pages and blocks

### DISCUSSION THREAD

**Discussion by jtrakk** [athensresearch/athens#711] [General]
[Roam and Athens: Boons and Pitfalls](https://elliotsayes.medium.com/roam-and-athens-boons-and-pitfalls-61cad7ab80fa) explains how having Pages in the model requires users to make an awkward choice between Pages and Blocks. The article proposes unifying pages and blocks into a single type and explains several advantages it would have. 

What do you think about this proposal?

**Comment by leogao2** [comment]
I personally like this proposal - I've run into issues with deciding which level of abstraction of concepts are to receive a page and which are merely a block (square bracket linking is more convenient, but blocks are more convenient to navigate within a single topic), and by merging blocks and pages, this would be a lot simpler (not to mention this abstraction feels more elegant).

**Comment by aaronweewee** [comment]
Remnote does this and it simplifies and make things streamlined and reduces one layer of friction, it also makes embedding a lot easier. But from what I can tell, Athens is following the steps of LogSeq and Roam, so I don't think they will do this

**Comment by elliotsayes** [comment]
Author of the article here :)

I also doubt that Athens will prioritize this until it has feature parity with Roam. But I would point out that the proposed *data model* behind it is technically compatible with an exact clone of Roam, but would also allow more options in the future. Something to consider @tangjeff0 :P

**Reply by nhanvu327** [reply]
I mean why the hell Athens must be the exact clone of Roam though.

**Reply by jefftangx** [reply]
Haha most of Roam features we don't want actually :)

**Reply by jtrakk** [reply]
@tangjeff0 Can you elaborate on that, or is it written up somewhere? I've been hoping to switch from Roam to Athens as soon as it's available for my distro, but I'd like to learn more about the intended design differences.

**Reply by jefftangx** [reply]
@jtrakk what features are mission-critical from Roam that Athens doesn't have?

**Reply by jtrakk** [reply]
@tangjeff0  My only experience with Athens is fiddling around in the demo app. 

The big things I notice missing in 2 minutes are:

- Paste image from clipboard. Roam uploads and displays it.
- `/file` upload.
- Sync status indicator. Google Docs, Roam, and Dropbox have a little indicator light that says whether my changes are synced yet. Communicating system status accurately is really important to having confidence that my work is being saved.
- A lightweight query system. `(a & b | c) | !f` would be fine.
- I sometimes get "parse error" tool-tips appearing in the demo after I mashed the keyboard and typed a bunch of apostrophes. I don't know what it's talking about but Roam doesn't do that.


I can't check easily in the demo but I'd want to make sure that it's robust:

- Roam block actions (create block, move block) are sluggish on a page with a few thousand blocks and it's terrible. 
- The search box should stay fast into millions of pages.
- Incremental deploy process with rollback. Deploy first to a small group, then incrementally increase. When you make some changes that break stuff, monitoring should be able to notice changes in the metrics and roll back before deploying to all users.
- Real-world latency monitoring. A few months ago Roam had an update that added a catastrophic-backtracking regex, and it made Roam unusably slow for several days -- not okay for users who consider it their second brain. See [how Chromium does real-world performance monitoring](https://blog.chromium.org/2021/04/digging-for-performance-gold.html).

**Reply by jefftangx** [reply]
A few of your listed bullets are available on Electron, not on web demo: sync indicator and image paste.

IMO the 3 most important features Roam popularized were bidirectional links, block references, on top of an outliner. Most of the features many users ask for like `/file` are not fundamentally new or a different paradigm, just a faster horse -- expected features for a notetaking app. The two bullets you mention that are interesting/different are a `lightweight query system` and `fast search into millions of pages`.

In the short-term, we should just improve the core to be very robust and production-grade, as you mention. Stability, no data loss, and performance are the most important principles for a knowledge tool. 

That said, there is still a ton of room for innovation at both the interface and data level that no one has really addressed yet. We've really only scratched the surface. The more robust we make this core, the broader and deeper Athens can be as a platform. That's what I'm most excited about!

**Comment by zhangxiubo** [comment]
I'd second this proposal.  From a purest point of view, a page can be modelled as a named block/node.

As an example, when a user types in `[[Athens]]`, a reference (e.g. `((abc123XYZ))`) to the block whose "alias" attribute contains the value `Athens` can be inserted in its place instead;  the UI then renders "Athens" as the text of the link.

**Reply by jtrakk** [reply]
In that implementation, what happens if two different blocks have `Alias:: Athens`?

**Reply by zhangxiubo** [reply]
Then the users are offered to choose which block they meant to reference (the same way we disambiguate block references).  The alias attribute is understood and parsed by the UI/app to generate the choices;  the actual linking is maintained by the block reference.

The problem I can see with this approach is when a block has multiple aliases --- the block reference itself can't distinguish which alias should be rendered as the label for the reference, so we might need a `[Athens]((abc123XYZ))` type syntax.

**Reply by elliotsayes** [reply]
I'm not really sure what this solves. You can comfortably use the [[Page link]] syntax to link to a particular text block in the notebook. In general, the ((blockid123)) linking is an ugly last resort.

Aliasing here is a seperate question... however I do think that aliases should be tied to the target text blocks themselves (DRY), and selected by the referencer.

**Reply by zhangxiubo** [reply]
> You can comfortably use the [[Page link]] syntax to link to a particular text block in the notebook.

I didn't know this is possible --- has it been implemented yet or it's more of a proposal?

Alias by itself is indeed another issue.  I am only bringing it up here to show how pages can be modelled as aliased block references when we implement pages as named blocks.

**Reply by elliotsayes** [reply]
No it is not possible currently, because the [[Page Link]] is reserved for pages. But if page links are merged with ((text blocks)) there is no reason not to use the formers syntax for the latter.

Remember that page is not the same as an block "alias", it is the literal block. I don't know why an alias should be necessary for a block, if you wish to link to the page 'Athens' simply make a text block called 'Athens' (marked as Page-type) and create any number of child text blocks with further details.

**Reply by jtrakk** [reply]
@elliotsayes If I type [[Athens]], what list should it look through to find a match: is it the text of blocks with `ispage :: true` attribute, or is it searching for a block with `title` attribute matching `:: Athens`?

**Reply by elliotsayes** [reply]
I'm thinking the first one :) Second one depends on how you want to implement aliases. Perhaps each text block could have alias list metadata, and could search through those also. Many possibilities

**Reply by zhangxiubo** [reply]
What I meant is that by giving a block an alias, it automatically implies the block is of a "page-type";  in other words, a "page" block is any block whose "alias list" contains at least one name.  In fact, if we follow this implementation, then there is not the concept of "page" anymore.

The alias (list) approach also addresses how "aliasing" can be handled in such a setting; a single block can be addressed in multiple ways.

**Reply by zhangxiubo** [reply]
As for whether the reference should be represented as block ID or simply using one of the names of the block --- if Athens' UI can guarantee the uniqueness of names then addressing by block name/alias would be preferred;  otherwise the UI automatically inserting the block-id-alias syntax might be necessary.

**Reply by elliotsayes** [reply]
I dont think it makes sense for pageness to be tied to aliasness. Remember that currently, a page is tied to a *list* of text blocks (its children).

What you are proposing is fundamentally different to the structure of Roam/Athens notebooks, regardless of linking mechanics.

**Reply by jtrakk** [reply]
> What you are proposing is fundamentally different to the structure of Roam/Athens notebooks

Could you expand on how it is different?

**Reply by zhangxiubo** [reply]
@elliotsayes I don't think this is fundamentally different.  Perhaps the use of "alias" in my descriptions is confusing.  Let me rephrase:

1. The only constructs in a graph/notebook are blocks.
    1. Each block can have child blocks.
    2. Unlike a tree of blocks, there is not a single "parent" block for any block.
1. Each block has a unique ID.
1. A block is _anonymous_ by default, i.e. it doesn't have names attached to it.
1. If a block has at least one name attached to it, then it becomes a _named block_.
    1. Each named block has a _primary name_, which is the first name in its list of names.
1. Named blocks behave like pages, e.g. 
    1. When the user "drill down" to the block by clicking the bullet in front of it, its primary name is displayed at the top, just like the title of a page.
    1. Named blocks can appear in the graph visualisation, whereas anonymous blocks do not.
    1. Users can address a named block using the `[[Name of the Block]]` syntax.
        1. If more than one named blocks match the query, the user will choose which block it meant to reference.
        2. Once chosen, an aliased block id is inserted, e.g. `[Name of the Block]((abc123XYZ))`.

In fact I believe this is quite similar to how Tiddlywiki organises its "tiddlers"/blocks;  the only difference is that the user has to manually come up with the ID for the "tiddlers"/blocks.

**Reply by elliotsayes** [reply]
**Currently**
1 Page : N (child) text blocks relationship

**Your proposal**
1 Page : 1 (aliases) text block relationship

**Reply by zhangxiubo** [reply]
@elliotsayes I see what you mean.  If we treat a page as a block whose content is the title, then they would be equivalent.

**Reply by elliotsayes** [reply]
But you're still coupling "pageness" with "aliasness", when they are really independent concerns. I don't see why text blocks need an independent name / alias to become a page, or why I wouldn't be able to alias a non-page text block.

**Reply by zhangxiubo** [reply]
In that sense yes you are right.  As for the aliasing --- my rationale is that there should be no reason to prefer one name over others, thus my proposal of a separate list of names attached to a block.

**Reply by elliotsayes** [reply]
Thanks @zhangxiubo , I think that's a very good idea on its own.

**Reply by zhangxiubo** [reply]
@elliotsayes Another concern I had with addressing text blocks directly with `[[Athens]]` type of syntax is what should happen when there are conflicts --- one can create a block with "Athens" as its content, mark it as a page block, and start to referencing it with `[[Athens]]`; the user may later create another block with the exact same content, wouldn't that lead to ambiguity?

I guess one option would be to replace any promoted competing text block by a embedded copy of the original page block when user "pagefy" a competing block.

**Reply by elliotsayes** [reply]
That's a great question, and I completely agree with your suggestion. This has the benefit of having a similar interface for disambiguating aliases with the same label that point to different blocks.

**Reply by zhangxiubo** [reply]
In that case if the user decides to promote a block which competes with another existing page block, then Athens needs to be able to merge two graphs; this could be a non-trivial task, if not impossible.

This actually circles me back to my original proposal, the only difference being that the content of a "page"/named block should be disabled, so the block only exists to serve as some sort of "meta-block" to which names and attributes can be attached.

An example named block could look like

```
block-id: abc123XYZ
names: ["Shannon's Entropy", "Information Entropy"]
content: null
```

When the user types `[[Information Entropy]]`, a query is triggered to retrieve all the blocks whose `names` attribute contains "Information Entropy"; the user then selects the desired block, and the reference `[Information Entropy]((abc123XYZ))` is inserted. 

The user may also drill down into the block.  The primary name ("Shannon's Entropy") is displayed at the top as its title.  No option is presented to edit the content of the named block, and the user can start adding child blocks the same way as editing a page.

**Reply by zhangxiubo** [reply]
To avoid having to deal with competing page blocks for a "is-page" based implementation, a page block may look like the following:

```
block-id: abc123XYZ
is-page: true
alias: ["Information Entropy"]
content: "Shannon's Entropy"
```

When the user types `[[Shannon's Entropy]]`, a query is triggered to retrieve all the blocks whose `is-page` attribute is `true` and whose content is "Shannon's Entropy".  Then the user selects the desired block, and the reference `((abc123XYZ))` is inserted.

I guess the key question here is: can we just insert `[[Shannon's Entropy]]` as is and use "Shannon's Entropy" as the reference key?

I think the answer is no, because otherwise we would be using the content of a block to resolve page references, and the content of a block is not meant to be unique like `block-id` is.

**Reply by zhangxiubo** [reply]
So in both cases, the `[[Name of a Page]]` syntax is really just a way to insert block references to special "page" blocks.  The `is-page` approach is perhaps more intuitive for the user;  the name list approach feels more consistent to me because it treats each name of the block equally.

**Reply by hftf** [reply]
I agree that pages and blocks should be unified, but it is also still important to retain a concept of "pageness": keeping things that I have consciously decided are pages (by flipping a boolean flag, or by giving them a name, etc.) distinct from all things I have written. One primary purpose of "pageness" is to have two link behaviors (such as `[[` vs. `((`) – one that searches and autocompletes only from the small set of pages (intentionally marked things), and one that does from all writings (unmarked).

**Reply by elliotsayes** [reply]
@hftf agreed that is important :) another option is to prioritise pages when searching, i.e. top 5 hits are always pages

@zhangxiubo sorry for late reply. theres a couple more possibilities here, both sound a little weird but tie into other changes. the benefit of these is that they avoid mandatory naming and/or exposing the ((blockid123)) syntax to the user

1) block id metadata is stored behind the scenes when using the context-aware autocomplete, and not shown to the user even in edit mode. an implication of this is that text blocks are no longer defined by their plain text string. 

2) block text must be unique. this might sound absolutely crazy, but relies on the assumption that the *same* text block object can appear in *multiple* places in the notebook hierarchy.

my preference is actually both :) but the rationale is too long winded to type here

**Reply by zhangxiubo** [reply]
@elliotsayes Option 1 is certainly viable; the underlying text string can still contain the block IDs, it's just that the UI features a semi-plain-text edit mode that resolves and renders styled versions of the links but keeps everything else in plain text.

As for Option 2, I agree that's one way of doing it, and I actually think allowing the same block object to appear in multiple places (i.e. block transclusion) is the "right" way of doing knowledge graphs (Athens and other Roam-likes are mostly "knowledge trees" with symbolic links at this stage).  My worry is that using text content as IDs may lead to unexpected/undesired behaviours, e.g. "Athens the city" vs. "Athens the software" ambiguity, and also changes to a page block propagates unexpectedly to all its citations.  Of course, users can make it work if they carefully disambiguate the blocks like how it is done by Wikipedia, but I wonder if this is expecting too much from the average users here.

**Reply by zhangxiubo** [reply]
As our discussion continues, I think it is beneficial to reiterate the original goal for unifying pages and blocks, as pointed out in @elliotsayes' blog, is to relieve the users from the anxiety of having to answer the question "Should I use a page or a block", and also the question "Is this the right place for this text block?" whenever they create a block.

IMHO, I feel the canonical solution to this problem would be to allow a block to exist independently without a parent, just like pages, hence the desire to unify them.  At a conceptual level, blocks need to be organised as nodes in a graph rather than nodes in a tree.

**Reply by elliotsayes** [reply]
Yes agree with everything you are saying, it seems we share the same vision of notebook structure and only vary on a few details. And you are right that I'm underselling the disambiguation problem, and that "block transclusion" as you eloquently put it is not a full solution by itself. Forcing disambiguation via raw block text e.g. "Athens (city)" is ugly and annoying to maintain as you say. 

Not exactly sure what the answer here is... maybe you are right that it should tie into some sort of first class aliasing system e.g. "Athens" => "Athens (city)" || "Athens (software)" or even definition system "Athens" => "Modern capital of [[Greece]]" || "Open source note taking software modelled after [[Roam Research]]".

Potentially it ties in to another idea I had, being able to tag text blocks with a "context" (just a native text block itself). In this case there would be an "Athens" in the context of "Geography" for example, and another "Athens" in the context of "Software". The assumption here being that a text block must be unique within a context.

Since I'm getting a bit off topic here, if you'd prefer to take this conversation offline feel free to msg me on discord at myn#0344 :-)

**Reply by jefftangx** [reply]
@elliotsayes I would love if this convo continued in public! Learning a lot. Can create a separate discussion too :)

---

re @zhangxiubo 

> IMHO, I feel the canonical solution to this problem would be to allow a block to exist independently without a parent, just like pages, hence the desire to unify them. At a conceptual level, blocks need to be organised as nodes in a graph rather than nodes in a tree.

I feel that this is the purpose of daily notes in some sense. To create a block where there isn't a conceptual parent, but a parent anchored in time. This is why this kind of tool is valuable for research, where you are gathering information and insight but you are not sure yet of the conceptual scaffolding.

**Reply by zhangxiubo** [reply]
re @tangjeff0 

I agree with your observation.  That's pretty much how I use Roam-likes nowadays: dump everything on daily notes and re-organise them as needed.  I feel this workflow works quite well for keeping track of project progresses if the user is comfortable following an opinionated system.

I think there are also cases where a less time-dependent approach is called for.  One example would be structured long-form writing.  It is still useful to evolve the thinking with daily notes of course, but the structures of such writing may not align well with the temporal order in which the author comes across the ideas.

**Reply by zhangxiubo** [reply]
As for the discussion on the unifying of pages and blocks: perhaps it could be useful to anchor the discussion with some concrete use cases.  Take the following note I come up with as an example:

<p align="center">
<img src="https://user-images.githubusercontent.com/2065923/117138989-9e2dac80-ada3-11eb-8961-afdbe6c09315.png" width="80%" />
</p>

It seemed pretty obvious to me in hindsight that the topic of "Unifying pages and blocks" should make into a page, but there is another voice in my head asking the question:  should I avoid starting a new page at all in the initial note (the one in May 03, 2021) and just making references to the original block instead? Like this:

<p align="center">
<img src="https://user-images.githubusercontent.com/2065923/117141285-417fc100-ada6-11eb-8d9c-d2119453f93d.png" width="80%" />
</p>

And this is where the "Should I use a page or a block?" question comes in.

As a programmer, I feel this block-only mental model is easier to maintain; another advantage is that I don't need to worry about disambiguating a page (e.g. one can refer to various different "Transclusion"s without conflating them all into one page).

What annoys me with this approach, however, is that there is always this special "genesis block" hanging around.  What I would like to see is for the block to somehow "promote" itself into a parentless page-block.  What this "promotion" entails exactly can be a subject for further debate, I guess.

I think both of these strategies can get the job done, and at the end of the day it's more of a personal preference which one way is better.

---
P.S. I really like how I can pull all the _relevant_ contexts of the discussions into one place just by opening the `[[Unifying pages and blocks]]` page.  To me this is where RR and Athens really shine --- other non-outliner/markdown-based tools, e.g. Obsidian, would have difficulty showing the right contexts due to the lack of built-in block structures.

**Reply by jtrakk** [reply]
I think what's going on is that

1. I write a bunch of notes about Athens throughout my graph, tagging each of them with #Athens.
2. I want to create a cohesive document where I can view and manipulate all of the notes about Athens.

Currently this workflow is awkward. The linked references are all at the bottom and I can't move them around. It needs a place where all of the blocks tagged #Athens appear. What happens if I add a new block to this document: does it automatically get tagged #Athens? I'm not sure.

Either way, that's not quite good enough either. I want to have a document containing all of the blocks either tagged #Athens or tagged #Roam, and which I can reorder and edit. Roam has `{{query:}}` but that interface is clunky. Now what happens if I add a new block to this document? I'm not sure.

Disclaimer: I use Roam.

**Reply by zhangxiubo** [reply]
re @jtrakk

I think that's a pretty tough requirement you were describing there.  One possible solution would be:

1. The UI allows you to execute a command, say `pull {{query:}}`, that pulls all the existing blocks tagged with `#Athens` at the time of its execution, and attach them as child blocks to a parent block `P` of your choice.
2. Say the blocks pulled are `A`, `B`, and `C`.
3. You can then edit/re-arrange the blocks as needed, say the structure now looks like:
    ```
    P
    |- N
    |- A
       |- C
    |- B
    ```
    where `N` is some comment block you added.
4. You compose some new blocks `D` tagged with `#Athens` elsewhere.
5. To update the document above, the same `pull` command can be re-executed at `P` to pull such new blocks to the end of `P`'s children, so we ends up with:
    ```
    P
    |- N
    |- A
       |- C
    |- B
    |- D
    ```
Roam has block embeds which may enable some function like this.  Another possibility is to have Athens/Roam implement block transclusions.

---
Edit: in fact, I think https://roamjs.com/docs/extensions/pull-references might be what you are looking for.

**Reply by jtrakk** [reply]
> in fact, I think https://roamjs.com/docs/extensions/pull-references might be what you are looking for.

That moves the blocks from the other page to this page. Block embed would be better. I'm not sure why it needs to be a separate step though. It could happen dynamically: a Page is just a block that embeds blocks tagged with the page's name.

**Reply by zhangxiubo** [reply]
@jtrakk You could be right that the process can be dynamic; it's just that I am not confident about having both manual and automated behaviours on the same page.

**Reply by jefftangx** [reply]
![image](https://user-images.githubusercontent.com/8952138/137197099-90322f52-b6c8-41d4-896b-71cee3b04422.png)

Bumping this thread because I've been thinking very hard about this (the meme above was definitely not inspired by my life at 2am last night).

Would also be down to setup a proper call with everyone sometime for this. This is super hard stuff! @zhangxiubo @jtrakk @elliotsayes 

Summary (apologies for any wrong accounting and not hitting all points):
- Forcing users to decide between blocks or pages adds friction.
- Blocks could _and perhaps should_ be allowed to exist without a page.
- Pages could nest within other pages, like a namespace or filesystem. This would avoid collisions. After all, blocks can nest within blocks.
- Page references and block references are presented to the user differently, which adds friction. Unifying them could lead to a single [[ notation. There could be additional affordances for sorting or filtering the results based on block attributes.
- ((uid1234)) is machine-readable, not human-readable.
- Currently, page references are unique and block references are non-unique. What would happen in the case of multiple blocks with the same text?
    - Block meta-data could be stored that the user cannot see in edit mode. Trade-off: you are searching based on meta-data, not the plain text.
    - Block text could be unique :exploding_head: 
- Asking users to disambiguate names Wikipedia style, e.g.  "Athens (the city)" vs "Athens (the application)" is too much work friction for average users.
- Pages are just "named" blocks. A block could have multiple "names" AKA "aliases."
- Anchoring unnamed blocks to daily notes and accreting them in a collection of references over time works, but it doesn't mean it's necessarily the best UX.

**Reply by zhangxiubo** [reply]
After much experimentation, I eventually settled with a solution based on Tiddlywiki; Tiddlywiki together with its awesome [Streams](https://github.com/saqimtiaz/streams) solved my needs pretty neatly;  definitely try it out if you haven't used it yet.  For the aliasing,  I used the [uni-link](https://wikilabs.github.io/editions/uni-link/) plugin.  The UX is probably not as streamlined as Athens/Logseq/RoamResearch, but I find it is good enough for day-to-day journaling.

**Reply by zhangxiubo** [reply]
The Streams plugin maintains a list of child tiddlers for each parent tiddler (a tiddler is the unified unit of content in Tiddlywiki).  Each tiddler also has a parent field, though it's optional.  Such data structures are used to implement a outliner in which each "block" is also a tiddler so there is no distinction between "block" and "page".

**Reply by zhangxiubo** [reply]
Disambiguation is a harder problem that I haven't solved yet.  My current strategy is to maintain Wikipedia-style disambiguated names for each topic.  There might be a better way though.

**Reply by elliotsayes** [reply]
@tangjeff0 what approach are you looking to take here? are you wanting to solve a couple of specific problems (instances of UX friction, homonyms)? Or redesign the app model around abstractions that better fit knowledge representation?
![Untitled 7](https://user-images.githubusercontent.com/7699058/139585162-380f1645-bedc-4d83-b6b6-d04ebf9422c7.png)

**Reply by jefftangx** [reply]
@elliotsayes why is it an either/or? We want to want to improve both knowledge representation and UX.

**Reply by elliotsayes** [reply]
It's a fundamentally different way of approaching the problem, and the solutions you end up with will have different tradeoffs. So far the work done in (accessible) PKMs like Athens is the yellow arrow, starting from a familiar abstraction. The problem we are discovering is that there are inherint issues in this model, that are addressed by layering on bandaids for each one. 

It's like trying to make a circle by starting with a square and repeatedly trimming the corners. You'll get closer to a circle, but to base your model on lines makes development tedious, and will never be perfect.

Taking the homonym example - this is a systemic problem in languages. The bandaid might be one of:
1. Namespacing system
2. Alias/disambiguation system

Either will be added complexity for the user, as well as a complication to the underlying app model. And expect to repeat adding complexity as new problems arise when scaling to, broader, deeper, highly interconnected and multiuser graphs.

The other way of approaching the problem would be starting from the left "unfamiliar abstraction" and building the pink arrow to make it more accessible. This way, the application model would perfect match the business problem (representing knowledge), and homonyms etc would only be treated as an interface problem, separate from the underlying model.

But that creates new problems: what is that "perfect model" (Hint: it may not involve text), and how do you bridge the gap ("simplify") back to something usable? Finally, by exposing more of that model to the user their underlying thinking will at some point have to "flick a switch" and adapt to its fundamental abstractions.

My gut feeling is that it could be too soon to try this approach, as there are still a fair amount of "complexity points" available before overwhelming the user, and PKMs haven't yet proven themselves as indispensable tools worth learning new abstractions for. On the other hand I see this decision as fundamental to the philosophical and marketing point-of-difference between Athens and Roam.

Thanks for reading my thesis lol

**Comment by jtrakk** [comment]
["The data model behind Notion's flexibility"](https://www.notion.so/blog/data-model-behind-notion) is an interesting explanation of how Notion uses blocks.

**Comment by solarkraft** [comment]
Agreed. Pages are just thin wrappers around blocks. You can already zoom into blocks and make them essentially like a page. 

Unifying the two concepts might also allow for scoping/nesting of pages, which relates to the namespacing issue I keep hitting a lot (e.g. an "Architecture" page for Athens, plus one about designing buildings). 

That said some kind of linkability/the UX around pages should be preserved; I found keeping all my notes in a large tree in Workflowy/Dynalist fairly inconvenient, because clicking through the tree structure is a bit cumbersome. I appreciate Athens' quick search feature.

**Reply by jefftangx** [reply]
How do you think the nesting of pages would look like for you @solarkraft ? I think I saw that LogSeq has this now

**Reply by solarkraft** [reply]
My current nesting approach is `[[[[Topic]] Sub-topic]]` (I was surprised it's possible!), before that it was `Topic: Sub-topic`.  Both are not very convenient to reference and fall apart after a single layer of nesting. 
My idea was that, like in programming, references to stuff in the current namespace could have the name omitted. 

I never thought of namespacing like nesting pages, but it sounds like it could be a good approach. 

But maybe all of this would be moot if the line between pages and blocks was gone, since I frequently find the choice between them difficult (and the choice is never really right). 

In the spirit of your list, here's my interpretation:
A page is a block in a global namespace that has some bold text on the top node and is treated a bit differently in search and referencing.
